### PR TITLE
Text file, newlines at end of files

### LINKED ISSUE #9802: We have zero-line text files that aren't empty
**Describe the bug**
Investigating #9798, I wanted to find the number of single line project files. In doing this I found some were reported as having 0 lines.

On review, @geekosaur is catching the mistake I'm often making of dropping or forgetting these newlines.

> Therefore, “lines” not ending in a newline character aren't considered actual lines. That's why some programs have problems processing the last line of a file if it isn't newline terminated.
>
> The advantage of following this convention is that all POSIX tools expect and use it.
> SOURCE: [Why should text files end with a newline?](https://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline)

I have a script to add newlines when missing and can use it to fix up some files that don't conform, for instance `*.hs` and `*.project` files. I didn't find any non-conforming `*.cabal` files. There are only two legitimate zero line (empty) projects;

- `./cabal-install/tests/fixtures/project-root/cabal.project`
- `./cabal-install/tests/fixtures/project-root/nix/cabal.project`

**To Reproduce**

```
$ find -name '*.project' -exec wc -l {} \; | sort -k 1
0 ./cabal-install/tests/fixtures/project-root/cabal.project
0 ./cabal-install/tests/fixtures/project-root/nix/cabal.project
0 ./cabal-testsuite/PackageTests/ExtraProgPath/cabal.project
0 ./cabal-testsuite/PackageTests/HaddockArgs/cabal.project
0 ./cabal-testsuite/PackageTests/HaddockWarn/cabal.project
0 ./cabal-testsuite/PackageTests/NewBuild/CmdRun/Script/cabal.project
0 ./cabal-testsuite/PackageTests/NewBuild/CmdRun/ScriptLiterate/cabal.project
0 ./cabal-testsuite/PackageTests/NewHaddock/Fails/cabal.project
0 ./cabal-testsuite/PackageTests/NewSdist/DeterministicTrivial/cabal.project
0 ./cabal-testsuite/PackageTests/PkgConfigParse/cabal.project
0 ./cabal-testsuite/PackageTests/Regression/T5782Diamond/cabal.project
0 ./cabal-testsuite/PackageTests/SDist/Respect-Project-File/cabal.project
0 ./cabal-testsuite/PackageTests/SDist/T5195and5349/cabal.project
0 ./cabal-testsuite/PackageTests/WarnEarlyOverwrite/cabal.project
...

$ cat ./cabal-testsuite/PackageTests/WarnEarlyOverwrite/cabal.project
packages: .⏎
```

**Expected behavior**
We add the missing newlines (I have a script for this).

```bash
#!/bin/bash
# SEE: https://unix.stackexchange.com/questions/31947/how-to-add-a-newline-to-the-end-of-a-file
for f in $(git grep -Il '' -- "$1"); do
    echo "$f";
    tail -c1 < "$f" | read -r _ || echo >> "$f";
done
```

Stretch goal: We stop relying on @geekosaur spotting these mistakes and add a check for this in CI.

### PR DESCRIPTION
**Pull request by philderbeast**
Fixes #9802 for `*.hs` and `*.project` files ~and adds a script taking a file glob pattern for doing this~, adding newlines at the ends of files where needed. Adds back the whitespace github workflow and uses `fix-whitespace` to fix non-conformance. Add a short note about this in the contributing readme.

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#other-conventions).

### CODE REVIEW

**Review by geekosaur — APPROVED** [review]

**Comment by ffaf1** [comment]
@andreasabel created `whitespace.yml`, let us see what he has to say.

**Comment by philderbeast** [comment]
@ulysses4ever I see you are a [fix-whitespace contributor](https://github.com/agda/fix-whitespace/graphs/contributors). Do you think we can get `fix-whitespace` and `fourmolu` to play nicely together?

File: doc/getting-started.rst
Comment (geekosaur) [inline doc/getting-started.rst:80]:
```suggestion
To run the executable, switch into the application directory with ``cd myapp`` and run
```

**Review by geekosaur — APPROVED** [review]

**Review by andreasabel — APPROVED** [review]
You could also use https://github.com/andreasabel/fix-whitespace-action:
```yaml
steps:
- uses: actions/checkout@v4
- uses: andreasabel/fix-whitespace-action@v1
```
This action downloads the pre-build binary of `fix-whitespace` and runs it, and also caches it.

**Comment by philderbeast** [comment]
Thanks @andreasabel `fix-whitespace-action` is way less configuration. I reverted one of the corrections and it shows violations in the output.

**Comment by philderbeast** [comment]
Label merge+no rebase is necessary when the pull request is from an organisation.

**Review by ulysses4ever — CHANGES_REQUESTED** [review]
Instead of creating a separate workflow, can we add it to the quick-jobs workflow?

Can you check that fix-whitespace is run with a flush that will show the actual violations?

**Comment by philderbeast** [comment]
> I reverted one of the corrections and it shows violations in the output.

@ulysses4ever I checked and it shows violations.

**Comment by ulysses4ever** [comment]
Very cool, thanks!

**Comment by philderbeast** [comment]
@ulysses4ever, funny thing about moving it to quick jobs is that it now doesn't run quite as quickly!

**Comment by philderbeast** [comment]
In terms of developer feedback, putting it alongside (at the end of) quickjobs is a big degradation.

**Comment by ulysses4ever** [comment]
Mm, good point. My intention was not having it run quickly, though, and rather I dislike proliferation of standalone workflows. But maybe it’d be good to get an actually quick feedback from this job…

We need to fix quick-jobs to only download binaries and not build anything with ghc!

**Review by ulysses4ever — APPROVED** [review]

**Comment by philderbeast** [comment]
I do think that jobs like this one and the docs should run near instantly otherwise the wait is annoying.

**Comment by philderbeast** [comment]
@ulysses4ever are you OK if I revert the last change, putting whitespace into its own workflow?

**Comment by ulysses4ever** [comment]
Yeah, up to you, approved. Thanks!

**Comment by philderbeast** [comment]
@ulysses4ever I see there is something wrong with the cache in [quick jobs](https://github.com/haskell/cabal/actions/runs/8294806206/job/22700513058?pr=9804);

> Post job cleanup.
> Warning: Path Validation Error: Path(s) specified in the action for caching do(es) not exist,
> hence no cache is being saved.

**Comment by ulysses4ever** [comment]
Interesting... Maybe with a new ticket with the continuous-integration label.

**Comment by philderbeast** [comment]
How much squashing should I be doing with this given that the advice is:

> If you push a fix of a whitespace violation, please do so in a _separate commit_.

**Comment by philderbeast** [comment]
Sorry @Mikolaj for the lack of squashing. I asked about squashing but forgot to come back to it. Noticed when I got the cabal.head email.

**Comment by Mikolaj** [comment]
Ouch. With no-rebase and no manual rebase nor squash just before merging, the 12 commits are strewn artistically across other PRs from the several last days. I guess we have to get used to git art. :D

**Comment by andreasabel** [comment]
For the sake of bisecting, a linear history is really a plus.  Can't we enforce it on this repo?  (On the Agda repo, I disabled merge commits.)

**Comment by ffaf1** [comment]
Some contributors — and most importantly, some employers — do not fancy `cabal` to have write access their repos ([e.g.](https://github.com/haskell/cabal/pull/8981#issuecomment-1597278210)).

At first we did some [manual cherry-picking](https://github.com/haskell/cabal/pull/9051), but then we gave in and [`merge+no rebase` was introduced](https://github.com/haskell/cabal/pull/9427).

**Comment by andreasabel** [comment]
Can mergify do the rebase by copying the contributor branch to a new branch on this repo first?  Can this be set up?  This might be a technical solution.

If a PR cannot be rebased then it should be rejected by default.  Otherwise, we'll not get a linear history. 

Contributors with write access could be encouraged to push to a branch here rather than on their own repo.

**Comment by ulysses4ever** [comment]
@mergify backport 3.12

**Comment by ulysses4ever** [comment]
@philderbeast 

>>   Can you check that fix-whitespace is run with a flag that will show the actual violations?

> I checked and it shows violations.

Apparently, we misunderstood each other. What I meant is that it should show not only the names of the files with whitespace issues but also the issues themselves. It could be done by the `verbose: true` option of the action: 

https://github.com/andreasabel/fix-whitespace-action/blob/2a7678e2f1319ae6e9595acb06b8cfb46e38af23/action.yml#L17-L20

**Pull request merged**
